### PR TITLE
ScheduledJob client 

### PR DESCRIPTION
**Pull request by soltysh**
@erictune SheduledJob part 2, based on #24970 so first two commits doesn't count. 
This is still WIP, but it's here so you know :)

``` release-note
Introducing ScheduledJobs as described in [the proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/scheduledjob.md) as part of `batch/v2alpha1` version (experimental feature).
```

[Analytics]()

### CODE REVIEW

File: pkg/apis/batch/validation/validation.go [hunk 1]
```diff
@@ -155,3 +157,65 @@ func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList
...
+		validValues := []string{string(batch.AllowConcurrent), string(batch.ForbidConcurrent), string(batch.ReplaceConcurrent)}
+		allErrs = append(allErrs, field.NotSupported(fldPath, *concurrencyPolicy, validValues))
+	}
+
+	return allErrs
+}
+
+func validateScheduleFormat(schedule string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	_, err := cron.Parse(schedule)
```
Comment (erictune) [inline pkg/apis/batch/validation/validation.go]: @lavalamp @bgrant0607 is there a limit to how complex of parser we want to include in validation?  At some point, is it up to the controller to react to errors in complex fields?   In this case, a third-party library is used to parse a cronspec, which is pretty complex.  Not a performance issue, I don't think, but maybe a dependency issue.

Comment (erictune) [inline pkg/apis/batch/validation/validation.go]: @soltysh I think adding godep for robfig/cron is fine in this PR.  the only thing I am unsure about is whether validation should call Parse, or whether that should be left to the controller to parse, and then raise an event or set a condition if there is an error in the format.

**Comment by erictune** [comment]
LGTM.

File: pkg/apis/batch/validation/validation.go (hunk 1 above)
Comment (soltysh) [inline pkg/apis/batch/validation/validation.go]: Personally, I'd prefer to know the cron format is wrong at validation and not when the controller will actually try to run it. Although a complex, it's one of those things I'd expect validation to take care.

**Comment by soltysh** [comment]
@erictune there are a few bit still missing in this PR, but hopefully they will be addressed first thing in the morning. Once that is done I'll work on another PR adding kubectl elements.

**Comment by soltysh** [comment]
For now I'd like to have the API part to land, there's some release-note-* labeling issue.

**Comment by soltysh** [comment]
@erictune this is still not quite finished. The biggest problem with this PR is that until now there was no multi-version client testing framework in place :cry: My approach in the last commit ([Enable multi-version tests in client tests](https://github.com/kubernetes/kubernetes/pull/25475/commits/812df208c4d8a3b192e6ca6bcaa6b97514686b13)) but that needs @caesarxuchao eyes. If this is good enough for now, I'll continue with additional necessary changes to finish it. ptal 

I need to step away for a few hours, but I'll be online later and I'll try to reach out to you and @caesarxuchao about it on slack.

**Comment by soltysh** [comment]
@erictune one more option is to split the testing part into a separate PR (including scheduled jobs, because that needs to be tested only in batch/v2alpha1). IMHO, that would be the best option, since with client part in place we can work in parallel, wdyt?

**Comment by erictune** [comment]
```
FAILED   ./hack/../hack/verify-godep-licenses.sh	17s
Verifying ./hack/../hack/verify-godeps.sh
Checking for 'Godeps/' changes against 'origin/master'
```

**Comment by erictune** [comment]
```
Your godep licenses file is out of date. Run hack/update-godep-licenses.sh --create-missing and commit the results.
```

**Comment by soltysh** [comment]
@erictune updated

**Comment by soltysh** [comment]
@k8s-bot test this github flake: #25375

**Comment by soltysh** [comment]
@erictune `Godeps/LICENCES` has changed, I've rebased the entire PR to make sure it's up to date. Can you re-add lgtm label. Hopefully over the weekend it can be safely merged.

**Comment by soltysh** [comment]
@k8s-bot test this github flake: #25539

**Comment by erictune** [comment]
```
./pkg/apis/batch/types.generated.go up to date.
--- ./pkg/apis/batch/v2alpha1/types.generated.go	2016-05-18 03:49:13.480272061 -0700
+++ ./pkg/apis/batch/v2alpha1/types.generated.go.original	2016-05-18 03:43:00.624299288 -0700
@@ -4963,7 +4963,7 @@ func (x codecSelfer1234) decSliceSchedul
 
 			yyrg1 := len(yyv1) > 0
 			yyv21 := yyv1
-			yyrl1, yyrt1 = z.DecInferLen(yyl1, z.DecBasicHandle().MaxInitLen, 1024)
+			yyrl1, yyrt1 = z.DecInferLen(yyl1, z.DecBasicHandle().MaxInitLen, 1000)
 			if yyrt1 {
 				if yyrl1 <= cap(yyv1) {
 					yyv1 = yyv1[:yyrl1]
./pkg/apis/batch/v2alpha1/types.generated.go was out of date. Please run hack/update-codecgen.sh. (If you're running locally, this was run for you already.)
./pkg/apis/apps/types.generated.go up to date.
./pkg/apis/apps/v1alpha1/types.generated.go up to date.
./pkg/apis/rbac/v1alpha1/types.generated.go up to date.
./pkg/apis/componentconfig/types.generated.go up to date.
FAILED   ./hack/../hack/verify-codecgen.sh	373s
```

**Comment by soltysh** [comment]
@erictune yup, I'm just updating.

**Comment by soltysh** [comment]
Rebased and regenerated. @erictune mind re-applying the label?

**Comment by philips** [comment]
Why aren't scheduled jobs a third party resource and external controller?

**Comment by soltysh** [comment]
@philips we got a lot of questions about that functionality and after talking with everyone involved we agreed it's one of those features that should be part of the core of the system.

**Comment by philips** [comment]
@soltysh Do you have a link to that discussion? I don't see it on the proposal https://github.com/kubernetes/kubernetes/pull/11980 and the proposal itself doesn't mention it afaics https://github.com/soltysh/kubernetes/blob/10de4c2d51d14c1efa9ae1f9914136960c081475/docs/proposals/scheduledjob.md I am just searching for "third" so probably just missing it.

**Comment by soltysh** [comment]
@philips the discussion was held over several issues both k8s and openshift, you'll need to follow the links from initial comments. But generally it all started from a proposal for ScheduledJobs, actually. From that point the current Jobs arose and ScheduledJobs is just a continuation of the topic (see https://github.com/kubernetes/kubernetes/pull/11746).

**Comment by philips** [comment]
@soltysh OK, could you perhaps summarize in a few sentences what about this wouldn't fit in the third party resource model? We have done TPM support and a number of prototypes on top of third party and I would love to understand how we avoid adding more and more types over time to the core API. Not urgent so I will make a note to ping everyone post v1.3 code freeze.

**Comment by philips** [comment]
@soltysh The main thing in the back of my mind is that expression of time in cron format is potentially a step back. For example systemd timers[1] introduced a number of different ways of expressing time that users may want and that aren't captured in the ScheduledJob

- Run $now, and then once every 24 hours after $now
- Run every 8 hours. Essentially sleep 8 hours between completions. In cron you need to say run on 0, 8, 16.

[1] https://wiki.archlinux.org/index.php/Systemd/Timers#Timer_units

**Comment by soltysh** [comment]
@philips the schedule format was initially influenced by https://github.com/mesos/chronos, we thought about using [ISO8601](https://en.wikipedia.org/wiki/ISO_8601) but after some investigation [I've realized](https://github.com/kubernetes/kubernetes/pull/11980#issuecomment-164446977) it's too limiting and we went with the cron as being the most popular. I don't see any problems with supporting multiple schedule formats with it. Thankfully ScheduledJob is still alpha feature so we're able to make some significant changes to the API based on users and our self's feedback

**Comment by erictune** [comment]
@philips @brendandburns 

Regarding format:
- definitely open to adding ISO 8601 and [Google cron schedule format](https://cloud.google.com/appengine/docs/python/config/cronref#schedule_format), as they appear to be distinguishable from each other.   I guess systemd format could be supported too, if it is distinguishable from the others.
- I interviewed a number of Chronos users about their cron needs, and none seemed to love ISO 8601.  

-----

Regarding ThirdParty status:

If @solysh an I can get it done, this is going to be Alpha in 1.3 due to momentum.

From there, I had assumed it would go to Beta.  But I guess another option could be to decide to move from alpha to third-party.

We are going to discuss criteria for something being third-party at a future community meeting (@sarahnovotny is trying to pick a date).  That would be a good time to bring up the idea of ScheduledJob being ThirdParty.

If someone were to produce a third-party version of scheduledJob, well before 1.4, and show that was comparably useful, that would be persuasive argument for the latter path.  I have been thinking about this path while working on the controller.  My impression is that it is going to be harder than some people think.  But definitely worth trying, so we can improve thirdParty.  Interested in hearing what your experience has been working with third party.

**Comment by philips** [comment]
@erictune Sounds great. I will make a note to follow-up with you in a few weeks about what makes schedule jobs hard to do in third party.

**Comment by soltysh** [comment]
Reapplying lgtm label after rebase.

**Comment by soltysh** [comment]
@k8s-bot test this github flake: #21484

**Pull request merged**